File: src/engine/adapters/kafka/catalog.ts
import type { KafkaJS } from '@confluentinc/kafka-javascript';
import { encodePath, type TreeNode } from '../../../shared/domain/tree';
import { mapError } from './errors';

// Internal topics/groups (consumer offsets, transaction state, ...) are Kafka-internal
// bookkeeping, not anything a user would browse.
function isInternal(name: string): boolean {
  return name.startsWith('__');
}

// caps.ts's doc-comment table reads "cluster -> topics, consumer groups" — both are root-level
// siblings, not nested under each topic (a consumer group can span many topics, or none of the
// ones currently browsed, so nesting it under one topic would misrepresent it).
export async function listRoot(admin: KafkaJS.Admin): Promise<TreeNode[]> {
  const [topicNodes, groupNodes] = await Promise.all([listTopics(admin), listGroups(admin)]);
  return [...topicNodes, ...groupNodes];
}

// P32 D9/F15.1: fetchTopicMetadata returns the topic array directly now, not `{ topics: [...] }`
// — the array's own elements (name, partitions[].partitionId/leader/replicas/isr) kept their
// kafkajs field names in this client's own compat layer, so only the outer shape changed.
async function listTopics(admin: KafkaJS.Admin): Promise<TreeNode[]> {
  let topics: KafkaJS.ITopicMetadata[];
  try {
    topics = await admin.fetchTopicMetadata();
  } catch (err) {
    throw mapError(err);
  }
  const nodes = topics
    .filter((t) => !isInternal(t.name))
    .map((t): TreeNode => {
      const count = t.partitions.length;
      return {
        kind: 'topic',
        name: t.name,
        path: encodePath([{ kind: 'topic', name: t.name }]),
        // P23 D3: a topic no longer expands in the tree — its partitions moved into the
        // definition view. `detail` keeps the count as the tree's at-a-glance summary.
        hasChildren: false,
        detail: `${count} partition${count === 1 ? '' : 's'}`,
      };
    });
  nodes.sort((a, b) => a.name.localeCompare(b.name));
  return nodes;
}

async function listGroups(admin: KafkaJS.Admin): Promise<TreeNode[]> {
  let groups: KafkaJS.GroupOverview[];
  try {
    ({ groups } = await admin.listGroups());
  } catch (err) {
    throw mapError(err);
  }
  const nodes = groups
    .filter((g) => !isInternal(g.groupId))
    .map(
      (g): TreeNode => ({
        kind: 'consumerGroup',
        name: g.groupId,
        path: encodePath([{ kind: 'consumerGroup', name: g.groupId }]),
        hasChildren: false,
      }),
    );
  nodes.sort((a, b) => a.name.localeCompare(b.name));
  return nodes;
}

export async function listPartitions(admin: KafkaJS.Admin, topic: string): Promise<TreeNode[]> {
  let topics: KafkaJS.ITopicMetadata[];
  try {
    topics = await admin.fetchTopicMetadata({ topics: [topic] });
  } catch (err) {
    throw mapError(err);
  }
  const found = topics.find((t) => t.name === topic);
  return (found?.partitions ?? [])
    .slice()
    .sort((a, b) => a.partitionId - b.partitionId)
    .map((p) => ({
      kind: 'partition' as const,
      name: String(p.partitionId),
      path: encodePath([
        { kind: 'topic', name: topic },
        { kind: 'partition', name: String(p.partitionId) },
      ]),
      hasChildren: false,
    }));
}

File: shell/internal/adapters/kafka/catalog.go
package kafka

import (
	"context"
	"sort"
	"strconv"
	"strings"

	"github.com/twmb/franz-go/pkg/kadm"

	"github.com/kirathecat/kira-studio/shell/internal/storage/model"
)

// abbreviateCount mirrors @shared/format's abbreviateCount (every SQL adapter's own local copy —
// postgres/catalog.go's is the original) closely enough for the tree's own partition-count badge.
// No shared Go home exists for it yet (§4.2), and a partition count reaching four digits on a
// real cluster is possible, so a plain strconv.Itoa is not equivalent.
var abbreviateUnits = []struct {
	threshold int
	suffix    string
}{
	{1_000_000_000_000, "T"},
	{1_000_000_000, "B"},
	{1_000_000, "M"},
	{1_000, "K"},
}

func abbreviateCount(n int) string {
	sign := ""
	abs := n
	if abs < 0 {
		sign = "-"
		abs = -abs
	}
	for _, u := range abbreviateUnits {
		if abs < u.threshold {
			continue
		}
		scaled := float64(abs) / float64(u.threshold)
		var text string
		if scaled < 10 {
			text = trimTrailingZero(scaled)
		} else {
			text = strconv.Itoa(int(scaled + 0.5))
		}
		return sign + text + u.suffix
	}
	return sign + strconv.Itoa(abs)
}

func trimTrailingZero(f float64) string {
	s := strconv.FormatFloat(f, 'f', 1, 64)
	if len(s) >= 2 && s[len(s)-2:] == ".0" {
		return s[:len(s)-2]
	}
	return s
}

// isInternalGroup mirrors catalog.ts's isInternal — kept for groups only. kadm.ListedGroup has no
// internal flag the broker itself supplies, unlike a topic's own IsInternal (P58e E10).
func isInternalGroup(name string) bool {
	return strings.HasPrefix(name, "__")
}

// listRoot is catalog.ts's listRoot: topics ∪ consumer groups, root-level siblings — a consumer
// group can span many topics, or none of the ones currently browsed, so nesting it under one
// topic would misrepresent it.
func listRoot(ctx context.Context, adm *kadm.Client) ([]model.TreeNode, error) {
	topics, err := listTopics(ctx, adm)
	if err != nil {
		return nil, err
	}
	groups, err := listGroups(ctx, adm)
	if err != nil {
		return nil, err
	}
	return append(topics, groups...), nil
}

// listTopics is catalog.ts's listTopics. kadm.TopicDetail.IsInternal is the broker's own answer
// (P58e E10), replacing the TypeScript's name.startsWith('__') heuristic for topics.
func listTopics(ctx context.Context, adm *kadm.Client) ([]model.TreeNode, error) {
	meta, err := adm.Metadata(ctx)
	if err != nil {
		return nil, mapError(err)
	}
	meta.Topics.FilterInternal()
	nodes := make([]model.TreeNode, 0, len(meta.Topics))
	for _, t := range meta.Topics {
		count := len(t.Partitions)
		plural := "s"
		if count == 1 {
			plural = ""
		}
		detail := abbreviateCount(count) + " partition" + plural
		nodes = append(nodes, model.TreeNode{
			Kind: "topic", Name: t.Topic,
			Path:        model.EncodePath([]model.PathSegment{{Kind: "topic", Name: t.Topic}}),
			HasChildren: false, // P23 D3 — a topic's partitions moved into the definition view
			Detail:      &detail,
		})
	}
	sort.Slice(nodes, func(i, j int) bool { return nodes[i].Name < nodes[j].Name })
	return nodes, nil
}

func listGroups(ctx context.Context, adm *kadm.Client) ([]model.TreeNode, error) {
	groups, err := adm.ListGroups(ctx)
	if err != nil {
		return nil, mapError(err)
	}
	nodes := make([]model.TreeNode, 0, len(groups))
	for _, g := range groups {
		if isInternalGroup(g.Group) {
			continue
		}
		nodes = append(nodes, model.TreeNode{
			Kind: "consumerGroup", Name: g.Group,
			Path:        model.EncodePath([]model.PathSegment{{Kind: "consumerGroup", Name: g.Group}}),
			HasChildren: false,
		})
	}
	sort.Slice(nodes, func(i, j int) bool { return nodes[i].Name < nodes[j].Name })
	return nodes, nil
}

// listPartitions is catalog.ts's listPartitions. Still called even though the tree no longer
// expands a topic (P23 D4): StreamView.vue's partition filter popover is a second, live caller,
// re-fetched every time it opens.
func listPartitions(ctx context.Context, adm *kadm.Client, topic string) ([]model.TreeNode, error) {
	meta, err := adm.Metadata(ctx, topic)
	if err != nil {
		return nil, mapError(err)
	}
	detail, ok := meta.Topics[topic]
	if !ok || detail.Err != nil {
		return []model.TreeNode{}, nil
	}
	partitions := detail.Partitions.Sorted()
	nodes := make([]model.TreeNode, len(partitions))
	for i, p := range partitions {
		name := strconv.FormatInt(int64(p.Partition), 10)
		nodes[i] = model.TreeNode{
			Kind: "partition", Name: name,
			Path: model.EncodePath([]model.PathSegment{
				{Kind: "topic", Name: topic}, {Kind: "partition", Name: name},
			}),
			HasChildren: false,
		}
	}
	return nodes, nil
}
